File: simpliRetro/src/main/java/com/trayis/simpliretro/ReactiveUtil.java
package com.trayis.simpliretro;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Completable;
import io.reactivex.CompletableOnSubscribe;
import io.reactivex.Flowable;
import io.reactivex.FlowableOnSubscribe;
import io.reactivex.Observable;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.Single;
import io.reactivex.SingleOnSubscribe;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class ReactiveUtil {

    public static <T> Observable<T> prepareObservable(ObservableOnSubscribe<T> subscribe) {
        return Observable.create(subscribe).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
    }

    public static <T> Observable<T> prepareObservable(Observable<T> observable) {
        return observable.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
    }

    public static <T> Single<T> prepareSingle(SingleOnSubscribe<T> subscribe) {
        return Single.create(subscribe).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
    }

    public static <T> Single<T> prepareSingle(Single<T> single) {
        return single.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
    }

    public static Completable prepareCompletable(CompletableOnSubscribe subscribe) {
        return Completable.create(subscribe).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
    }

    public static Completable prepareCompletable(Completable completable) {
        return completable.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
    }

    public static <T> Flowable<T> prepareFlowable(Flowable<T> flowable) {
        return flowable.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
    }

    public static <T> Flowable<T> prepareFlowable(FlowableOnSubscribe<T> subscribe) {
        return prepareFlowable(subscribe, BackpressureStrategy.BUFFER);
    }

    public static <T> Flowable<T> prepareFlowable(FlowableOnSubscribe<T> subscribe, BackpressureStrategy strategy) {
        return Flowable.create(subscribe, strategy).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread());
    }
}

File: simpliRetro/src/main/java/com/trayis/simpliretro/ReactiveUtil.kt
package com.trayis.simpliretro

import io.reactivex.*
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

object ReactiveUtil {

    fun <T> prepareObservable(subscribe: ObservableOnSubscribe<T>): Observable<T> {
        return Observable.create(subscribe).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }

    fun <T> prepareObservable(observable: Observable<T>): Observable<T> {
        return observable.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }

    fun <T> prepareSingle(subscribe: SingleOnSubscribe<T>): Single<T> {
        return Single.create(subscribe).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }

    fun <T> prepareSingle(single: Single<T>): Single<T> {
        return single.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }

    fun prepareCompletable(subscribe: CompletableOnSubscribe): Completable {
        return Completable.create(subscribe).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }

    fun prepareCompletable(completable: Completable): Completable {
        return completable.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }

    fun <T> prepareFlowable(flowable: Flowable<T>): Flowable<T> {
        return flowable.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }

    fun <T> prepareFlowable(subscribe: FlowableOnSubscribe<T>): Flowable<T> {
        return prepareFlowable(subscribe, BackpressureStrategy.BUFFER)
    }

    fun <T> prepareFlowable(subscribe: FlowableOnSubscribe<T>, strategy: BackpressureStrategy): Flowable<T> {
        return Flowable.create(subscribe, strategy).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }
}
